skrf/circuit.py

```diff
@@ -710,30 +628,20 @@ def _Xk(self, cnx_k: list[tuple]) -> np.ndarray:
         -------
         Xs : :class:`numpy.ndarray`
             shape `f x n x n`
+        
+        References
+        ----------
+        .. [#] P. Hallbjörner, Microw. Opt. Technol. Lett. 38, 99 (2003).
         """
-        Xnn = self._Xnn_k(cnx_k)  # shape: (nb_freq, nb_n)
-        Xmn = self._Xmn_k(cnx_k)  # shape: (nb_freq, nb_n)
-        # # for loop version
-        # Xs = []
-        # for (_Xnn, _Xmn) in zip(Xnn, Xmn):  # for all frequencies
-        #       # repeat Xmn along the lines
-        #     _X = np.tile(_Xmn, (len(_Xmn), 1))
-        #     _X[np.diag_indices(len(_Xmn))] = _Xnn
-        #     Xs.append(_X)
-
-        # return np.array(Xs) # shape : nb_freq, nb_n, nb_n
-
-        # vectorized version
-        nb_n = Xnn.shape[1]
-        Xs = np.tile(Xmn, (nb_n, 1, 1)).swapaxes(1, 0)
-        Xs[:, np.arange(nb_n), np.arange(nb_n)] = Xnn
-
-        return Xs # shape : nb_freq, nb_n, nb_n
-
-        # TEST : Could we use media.splitter() instead ? -> does not work
-        # _media = media.DefinedGammaZ0(frequency=self.frequency)
-        # Xs = _media.splitter(len(cnx_k), z0=self._cnx_z0(cnx_k))
-        # return Xs.s
+        
+        y0s = np.array([1/ntw.z0[:,ntw_port] for (ntw, ntw_port) in cnx_k]).T
+        y_k = y0s.sum(axis=1)
+
+        Xs = np.zeros((len(self.frequency), len(cnx_k), len(cnx_k)), dtype='complex')
+        
+        Xs = 2 *np.sqrt(np.einsum('ij,ik->ijk', y0s, y0s)) / y_k[:, None, None]       
+        np.einsum('kii->ki', Xs)[:] -= 1  # Sii
+        return Xs
```

Comment:
Since you've implemented `splitter`, eventually `_Xk` could use directly it. However, maybe a price of a loss of performances... Up to you.

Reply:
Indeed, the use of a splitter makes very clear the connection purpose. But I think here it is however interesting to spare the `media` and `splitter` instances. I also thought of a `splitter_s` helper function used by both `media` and `circuit`, but finally it is also interesting to see directly in each class the computation that is used.